### ISSUE TITLE
Ability to terminate a SAS session from within an ipynb?

### ISSUE THREAD

**Issue by ghost** [sassoftware/saspy#137]
Hi.  I love the saspy package.  It is a great addition to the SAS System.  I have searched the docs and can't find a way to terminate a SAS session from within an ipynb.  I found the article about disconnecting from a network and then reconnecting to a running SAS session.  That is very useful.   But, it would also be useful to be able to TERMINATE a connection to a remote SAS server from within an ipynb.  I am asking since we have seen instances where stopping the Python kernel, and closing the ipynb does NOT terminate the running remote SAS session on our Linux SAS GRID.

Are there any plans to add a new command to the saspy package that would allow us to terminate a remote SAS session?  Or, perhaps this already exists and I've just missed it?  ;-)

**Comment by tomweber-sas** [comment]
Hey thanks!, yes, you can terminate the SASsession ('sas' in the case below) with the _endsas() method:
```
sas._endsas() 
```
I am concerned about you not seeing your grid session terminate though.
Two thoughts on that.
1) just closing the notebook doesn't stop the python process (which is also the saspy process). But stopping the process via 'restart notebook' or file->close&halt should shut down the python process and saspy should end the SAS session (does _endsas() internally).
2) There is a fix at master for a case where if you have the ability to 'reconnect' then when I shut down the SAS session, it didn't go away be cause it was waiting for a possible reconnect. I think that bug is only fixed at master, not out in the latest pypi package. It's possible you are seeing this if you don't have that fix. If you use the _endsas() method and the Grid workspace server doesn't go away, I suspect this is the issue. You can tell if you have the reconnect capability set up by just trying the disconnect method; what it prints out will either be it worked, or you can't use it cuz it's not configured.
```
sas.disconnect()

# you should see one or the other otf these messages
'Succesfully disconnected. Be sure to have a valid network connection before submitting anything else.'
or
'This workspace server is not configured for reconnecting. Did not disconnect.'
```
Thanks!
Tom

**Comment by JTH-03-027** [comment]
Thanks.  I did some testing and, yes, we have the ability to use the saspy disconnect() method, then disconnect from a network, and then re-connect to the identical SAS GRID session after re-connecting to the network. That is quite handy.

After I opened my new issue, I found this old issue:
https://github.com/sassoftware/saspy/issues/131

and saw the _endsas() method.  I tried it.  I got the message that the SAS Session had been terminated, but, the task was still running on the Linux SAS GRID.

However, the good news is that after I terminated the Python kernel, and shut down the ipynb, this time, the SAS GRID Task *did* shut down.  It took several minutes, but, it did shut down as expected.

So, in summary:


  1.  We can use the disconnect() saspy method to disconnect from a running SAS session, disconnect from a WAN, connect to a wifi router, and then successfully resume executing saspy methods in the ipynb.  Very nice.
  2.  We tried using the _endsas() method.  We received an indication in the ipynb that the session had been ended, but, that task showed up as still running on our Linux SAS GRID.
  3.  Terminating the Python kernel and closing the ipynb DID cause the Linux SAS GRID task to shut down correctly (after a delay of a couple minutes).

It is still unexplained why the Linux SAS Grid did not shut down when we terminated the kernel and ended the ipynb last Friday?

Thanks for the fast response to our question.

John

**Comment by tomweber-sas** [comment]
@JTH-03-027 Hmm, I'm still not liking it taking minutes to shut down. Are you observing the workspace process directly, from a linux shell? Or from a monitoring or reporting part of SAS Grid? You were right on it w/ #131. That's the fix for really shutting down I mentioned. In that issue I think there was a delay in seeing the SAS session terminate too, but I think it wasn't directly from the OS seeing that. I wonder if there's a delay with the infrastructure.
Do you know what you have the reconnect time limit set to? If it's more than the few minute delay (I think the default is 60 min) then I'm still confused at to the delay. If it's a smaller window, like the delay, then it might be the issue if 131. Do you know if you have the code from master installed, or what's out in pypi? If you look at the java file itself (in the saspy code; saspy/java/pyiom/saspy2j.java), you can tell if you have the fix by looking about like 262 - you'll see 'else if (eol.contains("ENDSAS"))'
in that block you'll see 'lang._release();' and also 'cx.close();'. If right after that cx.close you see a block of code starting with 'if (reconnect)' then you have it, else you don't.
That's the thing, when I do _endsas() (whether you submit it or I submit it when python is cleaning up during shutdown, it tells IOM to shutdown and close the session. It really shouldn't take any time for that to happen. 

BTW, glad you like, and can use, the reconnect ability. Yeah, I think that's a cool feature too. Glad IOM had it for me to take advantage of!

**Comment by JTH-03-027** [comment]
Hi,

I am monitoring the status of the Linux SAS GRID process using the gsas -s command as in the bitmap below.

[cid:image004.jpg@01D3F776.963406F0]

The lag in shutting down may have something to do with us connecting to the Linux SAS GRID from a Python session running on an Amgen instance of a “datascience.com” session?

I am going to forward your response to our Linux SAS Team.  They have set up the saspy package in Linux so should be able to answer your questions about which version of saspy we have installed, and also your other questions about the timeout settings?  Based on the behavior we see, I suspect we do NOT have the latest “master” version that has the fix to the _endsas() saspy method?

I will let you know what they find out when they look into that specific jar file.

Thanks for the quick replies.

John

**Comment by JTH-03-027** [comment]
I think that “gsas” is an internal Amgen SAS macro.  It stands for “grid sas” and returns information about SAS GRID jobs.

It includes options to kill jobs etc.

I have attached the bitmap which was embedded in the last e-mail as a JPG file.  It shows the results of executing a “gsas -s all” which means “ Show me the status of all SAS GRID jobs assigned to my user account”.

I have passed your questions on to our SAS GRID experts.  Hopefully they will be able to answer them for you soon?

John

**Comment by tomweber-sas** [comment]
Just checking in. Any updates?
Thanks,
Tom

**Comment by JTH-03-027** [comment]
Hi Tom,

Sorry, but, I have been buried with other stuff, and this is a pretty low priority since we only have the saspy connection set up in our “DEV” environment.  I’m 99% sure that we have the “old, standard” install of saspy in our Linux ds.com environment, but, I have not checked to verify.

We expect that once we download and install the later version of the sapsy package that the issue with not being able to kill a SAS session from the ipynb will be solved?

Thanks for the follow-up.

Do you have any estimate when the fix you describe that is in the github repo will become part of the ‘default’ saspy package?

John

**Comment by tomweber-sas** [comment]
No problem, I understand. I've been off on other projects a bunch myself.
I do need to take what's at master and build a new pypi package (what you refer to a s default I believe). 
I need to merge in some things from another branch and run it through it's regression tests. I'll try to get that done next week. 
As for that code solving this. I expect, but I don't believe we've determined what's actually happening. The only case I'm aware of where there's a delay in shutting down is the case with the reconnect capability, and the fix for that is at master, not in the current pypi package. So, I'll say that I expect that's the issue and that this will fix it. So, we'll see :)
I'll post back here when I build that new pypi so you can just do 
```
pip install saspy
```
as opposed to 
```
pip install git+https://git@github.com/sassoftware/saspy.git
```
which (the second one) will do the same thing, but grab the code from master.

Thanks!
Tom

**Comment by tomweber-sas** [comment]
@JTH-03-027 , I've gone ahead and created the new 2.2.5 pip package. It contains the fix for the delayed shutdown due to the reconnect feature.
Tom

**Comment by tomweber-sas** [comment]
@JTH-03-027 This is showing @ghost which says your id was deleted. Not sure what's going on there.
But, I've had another issue which was the same as this, #150 which we have verified that the fix in 2.2.5/6 addresses. So, I have every expectation it will work for you too. 
I'm going to close this issue as it doesn't look like you're on here anymore. But, as always, just reopen it, or open a new one if you still have any problems!
Thanks,
Tom
